[2/N][ST deprecate][BE] Remove Replicate Tensor convert from DDP and PTD

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95450

No use is found for this ST/Replicated Tensor based DDP. As part of ShardedTensor migration, let's remove this logic. Trying to undo everything in https://github.com/pytorch/pytorch/pull/75753.

Commit:
[2/N][ST deprecate] Remove Replicate Tensor convert from DDP and PTD

[ghstack-poisoned]

Commit:
Update on "[2/N][ST deprecate] Remove Replicate Tensor convert from DDP and PTD"


No one is using this ST/Replicated Tensor based DDP. As part of ShardedTensor migration, let's remove this logic. Trying to undo everything in https://github.com/pytorch/pytorch/pull/75753.

[ghstack-poisoned]

fduwjj commented:
@pytorchbot rebase

Commit:
Update on "[2/N][ST deprecate][BE] Remove Replicate Tensor convert from DDP and PTD"


No use is found for this ST/Replicated Tensor based DDP. As part of ShardedTensor migration, let's remove this logic. Trying to undo everything in https://github.com/pytorch/pytorch/pull/75753.

[ghstack-poisoned]

Commit:
Update on "[2/N][ST deprecate][BE] Remove Replicate Tensor convert from DDP and PTD"


No use is found for this ST/Replicated Tensor based DDP. As part of ShardedTensor migration, let's remove this logic. Trying to undo everything in https://github.com/pytorch/pytorch/pull/75753.

[ghstack-poisoned]

rohan-varma commented on test/distributed/optim/test_zero_redundancy_optimizer.py:
```diff
@@ -1223,32 +1219,6 @@ def closure_ddp():
-        "parameters_as_bucket_view",
-        [False, True],
-    )
-    def test_zero_model_parallel(
```
shouldn't we just remove the replicated tensor enable / disable, but keep the test (i.e. still call `_test_zero_model_parallel`), instead of removing the test completely?

cc @awgu who originally wrote this test.

fduwjj replied:
Oh I might got the meaning of "Model Parallel" wrong here. I will change back.

rohan-varma commented on test/distributed/optim/test_zero_redundancy_optimizer.py:
```diff
@@ -1385,71 +1355,6 @@ def _test_ddp_zero_overlap(
-            else hook_with_zero_step_interleaved
-        )
-
-        # Disable DDP + ReplicatedTensor since ZeroRedundancyOptimizer
```
This test looks like it's just disabling replicated tensor and then running the logic, so shouldn't we just remove the disable replicated tensor part versus remove the entire test?

rohan-varma commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -1738,48 +1731,6 @@ def forward(self, x):
-        local_shards = [Shard(torch.randn(5, 10, device=device), local_shard_metadata)]
-        st = init_from_local_shards(local_shards, [10, 10])
-        m = MyModule(st)
-        with _ddp_replicated_tensor(False):
```
same here, shouldn't we just remove the replicated tensor disable, and ensure that DDP still ignores Shardedtensors?

fduwjj replied:
ok, since we are deprecating ST anyway. I thought this is something we can also remove.. I will only remove the replicated tensor part for now.

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -177,7 +176,6 @@ def forward(ctx, reducer, state_dict, *inputs):
 
     @staticmethod
     def backward(ctx, *grad_outputs):
-        state_dict = ctx.state_dict
```
good catch!

rohan-varma reviewed (commented):
Not sure if we're removing some tests that should still run (just without replicated tensor).

wanchaol reviewed (approved):
Looks like there're tests getting deleted accidentally like Rohan mentioned, please revert them back and let's only delete replicated tensor related logic. Otherwise looks good to me, thanks for deleting unused code!

Commit:
Update on "[2/N][ST deprecate][BE] Remove Replicate Tensor convert from DDP and PTD"


No use is found for this ST/Replicated Tensor based DDP. As part of ShardedTensor migration, let's remove this logic. Trying to undo everything in https://github.com/pytorch/pytorch/pull/75753.

[ghstack-poisoned]

Commit:
Update on "[2/N][ST deprecate][BE] Remove Replicate Tensor convert from DDP and PTD"


No use is found for this ST/Replicated Tensor based DDP. As part of ShardedTensor migration, let's remove this logic. Trying to undo everything in https://github.com/pytorch/pytorch/pull/75753.

[ghstack-poisoned]

Commit:
Update on "[2/N][ST deprecate][BE] Remove Replicate Tensor convert from DDP and PTD"


No use is found for this ST/Replicated Tensor based DDP. As part of ShardedTensor migration, let's remove this logic. Trying to undo everything in https://github.com/pytorch/pytorch/pull/75753.

[ghstack-poisoned]

fduwjj commented:
@pytorchbot rebase

Commit:
Update on "[2/N][ST deprecate][BE] Remove Replicate Tensor convert from DDP and PTD"


No use is found for this ST/Replicated Tensor based DDP. As part of ShardedTensor migration, let's remove this logic. Trying to undo everything in https://github.com/pytorch/pytorch/pull/75753.

[ghstack-poisoned]

fduwjj commented:
Two failed tests are:
1. test_access_by_keys within test/dynamo/test_modules.py. No distributed component is even used there. It's a OOM error.
2. vision_maskrcnn within dynamo/torchbench.py. This is only triggered when "ciflow/periodic" is added. No distributed components are used there.

So I think this PR is safe to merge. Feel free to revert if this we believe this PR is the culprit.

fduwjj commented:
@pytorchbot merge -f "Confirmed that failed tests are not related to this PR."
